torch/testing/_asserts.py

```diff
@@ -498,20 +468,65 @@ def _check_mapping(
             f"Missing keys in the actual mapping: {sorted(missing_keys)}\n"
             f"Additional keys in the actual mapping: {sorted(additional_keys)}\n"
         )
+
     for key in sorted(actual_keys):
         actual_t = actual[key]
         expected_t = expected[key]
 
-        exc = check_tensors(actual_t, expected_t)
+        exc = check_data(actual_t, expected_t)
         if exc:
             return _amend_error_message(exc, f"{{}}\n\n{_MAPPING_MSG_FMTSTR.format(key)}")
 
     return None
 
 
+def _maybe_to_tensor(input: Any) -> Optional[Tensor]:
+    """Maybe casts the input to a tensor.
+
+    :class:`~torch.Tensor`'s are returned without modification. If :mod:`numpy` is available, :class:`numpy.ndarray`'s
+    are cast with :func:`torch.from_numpy`. Everything else is casted with :func:`torch.tensor`.
+
+    Args:
+        input (Any): Input to be cast.
+
+    Returns:
+        Optional[Tensor]: :class:`~torch.Tensor` if it :attr:`input` is castable and ``None`` otherwise.
+    """
+    if isinstance(input, torch.Tensor):
+        return input
+    elif _NUMPY_AVAILABLE and isinstance(input, np.ndarray):
+        return torch.from_numpy(input)
+
+    with contextlib.suppress(Exception):
+        return torch.tensor(input)
+
+    return None
+
+
+def _cast_inputs(check_tensors):
+    """Decorator that casts the two first positional inputs to :class:`~torch.Tensor`'s.
+
+    Casting is performed by :func:`_maybe_to_tensor`. If the casting of any input failed, returns a :class:`UsageError`
+    instead of executing the decorated function.
+    """
+
+    def wrapper(actual: Any, expected: Any, **kwargs):
+        a_t = _maybe_to_tensor(actual)
+        b_t = _maybe_to_tensor(expected)
+        if a_t is None or b_t is None:
+            return UsageError(
+                f"Both inputs have to be tensors or something that can be cast to one, "
```

Comment:
Not cast to a tensor but used to construct one from torch.tensor(), right?

Reply:
Or rather `torch.as_tensor` that avoids copies for certain types of inputs (tensors, ndarrays, other array-like objects implementing Array Interface, etc)?